Enable forward/backward compatibility in TS mobile

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56079 Enable forward/backward compatibility in TS mobile**
* #56652 Fix bug in emitUse to drop all values that are marked as drop

This PR processes the operators in TS to figure out if there is any argument that can be ignored for mobile interpreter. We ignore the argument if and only if it is positioned after all positional arguments and all invocations of the operator use the default value for this argument. 
Differential Revision: [D27828149](https://our.internmc.facebook.com/intern/diff/D27828149)

tugsbayasgalan committed:
[WIP] enable forward/backward compatibility in TS mobile

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] enable forward/backward compatibility in TS mobile"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] enable forward/backward compatibility in TS mobile"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] enable forward/backward compatibility in TS mobile"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] enable forward/backward compatibility in TS mobile"




[ghstack-poisoned]

iseeyuan commented on test/cpp/jit/test_interpreter.cpp:
```diff
@@ -143,6 +144,23 @@ TEST(InterpreterTest, Basic_CUDA) {
+
+  int hidden_size = 2 * input_size;
+
+  auto input1 = at::randn({seq_len, batch_size, input_size}, at::kCPU);
```
Are those inputs used?

tugsbayasgalan replied:
good catch

iseeyuan commented on test/cpp/jit/test_utils.cpp:
```diff
@@ -93,6 +93,25 @@ std::shared_ptr<Graph> build_lstm() {
+      %20 : int = prim::Constant[value=0]()
+      %21 : int = prim::Constant[value=9223372036854775807]()
+      %22 : str = prim::Constant[value="value"]()
+      %3 : Tensor  = aten::slice(%0, %1, %20, %2, %1)
```
Would be nice to put comments of schemas for aten::slice.Tensor and aten::slice.str

tugsbayasgalan replied:
sure

iseeyuan commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -477,6 +479,9 @@ struct CodeImpl {
   // map from unique of nodes to register in register table
   std::unordered_map<Value*, int> value_to_reg_;
 
+  // map from operator name to trailing unnecessary inputs
```
Could you add an example in the comment? 
```
// Example: for a schema of aten::foo.str
// aten::foo.str(arg0: str="default", arg1: int=0, arg2: bool=False, arg3: float=0.0)
// If the usages in a graph is
// aten::foo("somestr", arg1=0, arg2=False, arg3=0.0)
// aten::foo("somestr", arg1=1, arg2=False, arg3=0.0)
// op_to_unnecessary_args_["aten::foo.str"] = ?
```

tugsbayasgalan replied:
sure

iseeyuan commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -634,11 +715,35 @@ struct CodeImpl {
+
     if (op.hasOperation() && op.schema().is_vararg()) {
-      insertInstruction(OPN, operator_table_.size(), node->inputs().size());
+      insertInstruction(OPN, operator_table_.size(), necessary_size);
```
@gmagogsfm , is it safe to update vararg operators (OPN)?

tugsbayasgalan replied:
For vararg operators, our algorithm would return 0 because user would always give inputs. (e.g there won't be any unnecessary arg in the end). Therefore I think the necessary_size will always be same node->inputs.size(). Maybe I shouldn't even change this?

gmagogsfm replied:
Yeah, I agree with you Tugsu that this probably shouldn't be changed.

iseeyuan commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -634,11 +715,35 @@ struct CodeImpl {
   void emitOperator(Node* node) {
-    emitLoadInputs(node->inputs());
     const Operator& op = node->getOperator();
+    auto unique_op_name =
```
Could  we have unit test to cover this part?

iseeyuan reviewed (commented):
Thanks! Could you add summary to this PR?

tugsbayasgalan committed:
Update on "[WIP] enable forward/backward compatibility in TS mobile"


Differential Revision: [D27828149](https://our.internmc.facebook.com/intern/diff/D27828149)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -115,7 +115,10 @@ std::pair<IValue, c10::optional<IValue>> getFunctionTuple(
   Inline(*graph);
 
+  // TODO: to use it for mobile, you need to enable it here.
+  // torch::jit::Code code(graph, func.name(), true);
```
To enable it, we can consider making a subclass of `Code` that's for mobile only.

gmagogsfm commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -634,11 +726,36 @@ struct CodeImpl {
   void emitOperator(Node* node) {
-    emitLoadInputs(node->inputs());
     const Operator& op = node->getOperator();
+    auto unique_op_name = op.schema().overload_name() != ""
```
Though the overhead added here is low, ideally we should still avoid it for normal TorchScript execution. One way to do so is to subclass interpreter for mobile export. The subclass should override emitOperator with an implementation with additional logic about ignoring arguments. This way, normal TorchScript execution would not be slowed down at all. What do you think?

gmagogsfm commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -520,6 +540,74 @@ struct CodeImpl {
     return constant_table_;
   }
 
+  void process_ops_for_mobile() {
```
IMHO, this algorithm can be optimized in implementation. What we really need in the metadata table is a mapping from op type to number of arguments serialized. So we can implement it this way to achieve equivalent functionality as what's currently implemented:

Python-like pseudo code
```
op_to_num_args = {}
for op in graph:
    if op not in op_to_num_args:
           op_to_num_args[op] = 0

    for (i, input) in enumerate(op.args()):
        if input is necessary:
            op_to_num_args[op] = max(i + 1, op_to_num_args[op])

    return op_to_num_args
```

Make sure to use iterators more when implementing in C++ to save map lookups.

tugsbayasgalan replied:
sure. makes sense

tugsbayasgalan commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -544,6 +632,10 @@ struct CodeImpl {
     return instructions_;
   }
 
+  const std::unordered_map<std::string, int> op_to_unnecessary_args() const {
```
sure

tugsbayasgalan commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -520,6 +540,74 @@ struct CodeImpl {
+        necessary_mark[i] = true;
+      } else {
+        auto schema_value = schema_args.at(i).default_value().value();
+        auto actual_value = toIValue(actual_inputs[i]);
```
sure

tugsbayasgalan committed:
Update on "Enable forward/backward compatibility in TS mobile"


This PR processes the operators in TS to figure out if there is any argument that can be ignored for mobile interpreter. We ignore the argument if and only if it is positioned after all positional arguments and all invocations of the operator use the default value for this argument. 
Differential Revision: [D27828149](https://our.internmc.facebook.com/intern/diff/D27828149)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Enable forward/backward compatibility in TS mobile"


This PR processes the operators in TS to figure out if there is any argument that can be ignored for mobile interpreter. We ignore the argument if and only if it is positioned after all positional arguments and all invocations of the operator use the default value for this argument. 
Differential Revision: [D27828149](https://our.internmc.facebook.com/intern/diff/D27828149)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Enable forward/backward compatibility in TS mobile"


This PR processes the operators in TS to figure out if there is any argument that can be ignored for mobile interpreter. We ignore the argument if and only if it is positioned after all positional arguments and all invocations of the operator use the default value for this argument. 
Differential Revision: [D27828149](https://our.internmc.facebook.com/intern/diff/D27828149)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -520,6 +540,55 @@ struct CodeImpl {
+    while (node) {
+      if (node->maybeOperator()) {
+        auto op_schema = node->getOperator().schema();
+        auto numIgnore = calculate_trailing_unnecessary_args(
```
```suggestion
        auto num_ignore = calculate_trailing_unnecessary_args(
```

gmagogsfm commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -520,6 +540,55 @@ struct CodeImpl {
+            ? op_schema.name() + "." + op_schema.overload_name()
+            : op_schema.name();
+        auto it = op_to_num_unnecessary_args_.insert(
+            std::pair<std::string, int>(unique_name, INT_MAX));
```
```suggestion
            std::pair<std::string, int>(unique_name, std::numeric_limits<int>::max()));
```

gmagogsfm commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -520,6 +540,55 @@ struct CodeImpl {
+
+    // keeps track of trailing unnecessary args
+    int trail_count = 0;
+    for (size_t i = 0; i < schema_args.size(); i++) {
```
Logic in this for loop can be simplified if we count backwards I think.

iseeyuan commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -477,6 +479,18 @@ struct CodeImpl {
+  // If the usages in a graph is:
+  //    aten::foo("somestr", arg1=0, arg2=False, arg3=0.0)
+  //    aten::foo("somestr", arg1=1, arg2=False, arg3=0.0)
+  // op_to_num_unnecessary_args_["aten::foo.str"] = 2
```
If we have another usage,
```
aten::foo("somestr", arg1=1, arg2=True, arg3=0.0)
op_to_num_unnecessary_args_["aten::foo.str"] = 1
```
Is that correct? If so, I'd suggest to use 
```
op_to_num_specified_args_["aten::foo.str"] = 3
```
The reason is, if later we update aten::foo.str and add another default arg, 
```
  // aten::foo.str(arg0: str="default", arg1: int=0,
  //               arg2: bool=False, arg3: float=0.0, arg4: bool=False)
```
In such case, the above 
```
op_to_num_unnecessary_args_["aten::foo.str"] = 1
```
in the old model cannot be used for backward compatibility. In the old runtime, it means the last argument's value, "0.0", is not specified. However, in the new runtime, it means the last argument's value, "Flase" is not specified. There is ambiguity. 
If we use 
```
op_to_num_specified_args_["aten::foo.str"] = 3
```
In old runtime we push "0.0" into the stack. In new runtime we push both "0.0" and "False" into the stack. Backward compatibility can be handled.

tugsbayasgalan replied:
makes sense. What should the map return in following case:

**schema:** 
prim::Foo(...) -> ...
**usages:** 
1. prim::Foo(arg1, arg2)
2. prim::Foo(arg1, arg2, arg3) 

Previously, we know the number of unnecessary args should be 0. But in this case, it is unclear. Maybe mobile team has to check if the schema is vararg as a special case handling? @gmagogsfm @iseeyuan

iseeyuan replied:
@tugsbayasgalan As discussed offline, we don't allow default arguments in vararg ops.From mobile side we could skip vararg ops on default args.

iseeyuan reviewed (commented):
Thanks for the updates! 
From my side, the only major issue is that we should use 
`op_to_num_specified_args_["aten::foo.str"] = 3`
instead of 
`op_to_num_unnecessary_args_["aten::foo.str"] = 1`
Please refer to the inline comment for details.

tugsbayasgalan committed:
Update on "Enable forward/backward compatibility in TS mobile"


This PR processes the operators in TS to figure out if there is any argument that can be ignored for mobile interpreter. We ignore the argument if and only if it is positioned after all positional arguments and all invocations of the operator use the default value for this argument. 
Differential Revision: [D27828149](https://our.internmc.facebook.com/intern/diff/D27828149)

[ghstack-poisoned]

iseeyuan commented on test/cpp/jit/test_interpreter.cpp:
```diff
@@ -143,6 +144,28 @@ TEST(InterpreterTest, Basic_CUDA) {
+  auto graph = build_mobile_export_analysis_graph();
+  MobileCode function(graph, "");
+  auto op_to_specified_args = function.op_to_num_specified_args();
+  ASSERT_TRUE(op_to_specified_args["aten::slice.Tensor"] == 4);
```
```suggestion
 ASSERT_TRUE(op_to_specified_args.size() == 2);
  ASSERT_TRUE(op_to_specified_args["aten::slice.Tensor"] == 4);
```

tugsbayasgalan replied:
done

iseeyuan commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -477,6 +478,18 @@ struct CodeImpl {
+  //               arg2: bool=False, arg3: float=0.0)
+  // If the usages in a graph is:
+  //    aten::foo("somestr", arg1=0, arg2=False, arg3=0.0)
+  //    aten::foo("somestr", arg1=1, arg2=False, arg3=0.0)
+  // op_to_num_specified_args_["aten::foo.str"] = 2
+  // This is because for all usages, last two args are not used. (e.g. same
+  // value as default schema value)
```
Number "2" here is symmetric, can be either specified or unspecified arg#. Suggest to change it to 3.
```suggestion
  //    aten::foo("somestr", arg1=1, arg2=True, arg3=0.0)
  // op_to_num_specified_args_["aten::foo.str"] = 3
  // This is because for all usages, at most first three args are specified. last one args are not used. (e.g. same
  // value as default schema value)
```

tugsbayasgalan replied:
done

iseeyuan reviewed (approved):
LGTM. Thanks!
@gmagogsfm could you also take a look?

tugsbayasgalan committed:
Update on "Enable forward/backward compatibility in TS mobile"


This PR processes the operators in TS to figure out if there is any argument that can be ignored for mobile interpreter. We ignore the argument if and only if it is positioned after all positional arguments and all invocations of the operator use the default value for this argument. 
Differential Revision: [D27828149](https://our.internmc.facebook.com/intern/diff/D27828149)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Enable forward/backward compatibility in TS mobile"


This PR processes the operators in TS to figure out if there is any argument that can be ignored for mobile interpreter. We ignore the argument if and only if it is positioned after all positional arguments and all invocations of the operator use the default value for this argument. 
Differential Revision: [D27828149](https://our.internmc.facebook.com/intern/diff/D27828149)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1048,6 +1082,93 @@ struct CodeImpl {
+  }
+
+  int calculate_necessary_args(
+      std::vector<Argument> schema_args,
```
Pass const reference instead of value

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -544,6 +562,10 @@ struct CodeImpl {
     return instructions_;
   }
 
+  const std::unordered_map<std::string, int> op_to_num_specified_args() const {
```
Consider passing reference back instead

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -508,7 +520,13 @@ struct CodeImpl {
-    // std::cout << *graph_ << "\n";
+  }
+
+  // since subclass of CodeImpl needs to populate
+  // op_to_num_specified_args, we seperate the calls
+  // that changes internals of CodeImpl into a separate
+  // function. User must manually call run() before using it
+  void run() {
```
We should probably avoid this because the logic here used to be part of constructor, now users of Code need to call it manually. 

Could you try marking `run` as virtual, override it in `MobileCodeImpl` and call it as last step inside constructor? This would avoid an additional manual call to `run`

https://www.programiz.com/cpp-programming/virtual-functions

tugsbayasgalan replied:
done

gmagogsfm reviewed (approved):
Looks great! Just some implementation nits.

iseeyuan commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -115,7 +115,9 @@ std::pair<IValue, c10::optional<IValue>> getFunctionTuple(
   Inline(*graph);
 
-  torch::jit::Code code(graph, func.name());
+  // TODO: to use it for mobile, you need to enable it here.
```
Is this statement up to date? I'm assuming I should use MobileCode?

tugsbayasgalan replied:
Yes sorry the TODO should have been removed. Basically emitOperator method in MobileCode is disabled right now sych that it calls emitOperator from its' parent class which is Code.

tugsbayasgalan committed:
Update on "Enable forward/backward compatibility in TS mobile"


This PR processes the operators in TS to figure out if there is any argument that can be ignored for mobile interpreter. We ignore the argument if and only if it is positioned after all positional arguments and all invocations of the operator use the default value for this argument. 
Differential Revision: [D27828149](https://our.internmc.facebook.com/intern/diff/D27828149)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Enable forward/backward compatibility in TS mobile"


This PR processes the operators in TS to figure out if there is any argument that can be ignored for mobile interpreter. We ignore the argument if and only if it is positioned after all positional arguments and all invocations of the operator use the default value for this argument. 
Differential Revision: [D27828149](https://our.internmc.facebook.com/intern/diff/D27828149)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Enable forward/backward compatibility in TS mobile"


This PR processes the operators in TS to figure out if there is any argument that can be ignored for mobile interpreter. We ignore the argument if and only if it is positioned after all positional arguments and all invocations of the operator use the default value for this argument. 
Differential Revision: [D27828149](https://our.internmc.facebook.com/intern/diff/D27828149)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1050,6 +1088,101 @@ struct CodeImpl {
+    run();
+  }
+
+  void run() override {
```
```suggestion
  void emitInstruction() override {
```

tugsbayasgalan committed:
Update on "Enable forward/backward compatibility in TS mobile"


This PR processes the operators in TS to figure out if there is any argument that can be ignored for mobile interpreter. We ignore the argument if and only if it is positioned after all positional arguments and all invocations of the operator use the default value for this argument. 
Differential Revision: [D27828149](https://our.internmc.facebook.com/intern/diff/D27828149)

[ghstack-poisoned]

iseeyuan reviewed (approved):
LGTM. Thanks!

tugsbayasgalan committed:
Update on "Enable forward/backward compatibility in TS mobile"


This PR processes the operators in TS to figure out if there is any argument that can be ignored for mobile interpreter. We ignore the argument if and only if it is positioned after all positional arguments and all invocations of the operator use the default value for this argument. 
Differential Revision: [D27828149](https://our.internmc.facebook.com/intern/diff/D27828149)

[ghstack-poisoned]
